zookeeper: 3.4.12 -> 3.6.2 & assorted changes

ztzg opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

  * [Apache ZooKeeper 3.4 End-of-Life 1st June, 2020](https://mail-archives.apache.org/mod_mbox/zookeeper-user/202004.mbox/%3C41A7EC67-D8F4-4C3A-B2DB-C2741C2EECA3%40apache.org%3E)
  * [Apache ZooKeeper 3.6.2 released](https://mail-archives.apache.org/mod_mbox/zookeeper-user/202009.mbox/%3CCACcefgfE-iDmmq3s7r1oRbvg1jY9QPF%3D3%3D2se-4SrEPtUqULng%40mail.gmail.com%3E)
  * [Apache ZooKeeper 3.6.2 is the latest stable release](https://zookeeper.apache.org/releases.html)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` (actually: `nixpkgs-review rev 67b116cd3b6`)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after) (minimal; smaller for main package.)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ztzg committed:
zookeeper: 3.4.12 -> 3.6.2

A big jump, but the structure hasn't changed much.

This recipe is still based on a binary release provided by upstream.

(It might be interesting to start doing our own builds at some point,
to split client from server, and/or to create packages for removed
"contribs" such as 'zooInspector'.  Upstream intends to further slim
down its release tarballs as most deployments only need specific assets.)

ztzg committed:
maintainers: add ztzg

ztzg committed:
nixos/zookeeper: adapt to zookeeper 3.6.2

This patch:

  * Removes an invalid/useless classpath element;
  * Removes an unnecessary environment variable;
  * Creates the required '/version-2' data subdirectory;
  * Redirects audit logging to the "console" (systemd) by default.

ztzg committed:
zookeeper_mt: 3.4.12 -> 3.6.2

ztzg committed:
zkfuse: adapt to zookeeper_mt 3.6.2

ztzg committed:
perlPackages.NetZooKeeper: adapt to zookeeper_mt 3.6.2

SuperSandro2000 commented on pkgs/development/libraries/zookeeper_mt/default.nix:
```diff
@@ -1,21 +1,64 @@
-  setSourceRoot = "export sourceRoot=${zookeeper.name}/src/c";
+  buildInputs = [
+    openssl
+    pkgconfig
```
```suggestion
    pkg-config
```

ztzg replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/zookeeper_mt/default.nix:
```diff
@@ -1,21 +1,64 @@
+, jre
+, libtool
+, openssl
+, pkgconfig
```
```suggestion
, pkg-config
```

ztzg replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/zookeeper_mt/default.nix:
```diff
@@ -1,21 +1,64 @@
+  sourceRoot = "apache-${zookeeper.pname}-${version}/zookeeper-client/zookeeper-client-c";
+
+  nativeBuildInputs = [
+    autoconf
+    automake
+    autoreconfHook
+    gettext
+    jre
+    libtool
```
```suggestion
    autoreconfHook
    jre
```

The hook should already contain them.

ztzg replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/zookeeper_mt/default.nix:
```diff
@@ -1,21 +1,64 @@
+{ stdenv
+, fetchurl
+, autoreconfHook
+, autoconf
+, automake
+, gettext
+, jre
+, libtool
```
```suggestion
, jre
```

ztzg replied:
Done.

SuperSandro2000 commented on pkgs/development/libraries/zookeeper_mt/default.nix:
```diff
@@ -1,21 +1,64 @@
     description = "Apache Zookeeper";
     license = licenses.asl20;
-    maintainers = [ maintainers.boothead ];
+    maintainers = [ maintainers.boothead maintainers.ztzg ];
```
```suggestion
    maintainers = with maintainers; [ boothead ztzg ];
```

ztzg replied:
Done.

SuperSandro2000 commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -15593,12 +15593,22 @@ let
+    NIX_CFLAGS_LINK = "-lzookeeper_mt";
     meta = {
-      maintainers = [ maintainers.limeytexan ];
+      maintainers = [ maintainers.limeytexan maintainers.ztzg ];
```
```suggestion
      maintainers = with maintainers; [ limeytexan ztzg ];
```

ztzg replied:
Done.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 104889` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>haskellPackages.hzk</li>
    <li>perl530Packages.NetZooKeeper</li>
    <li>perl532Packages.NetZooKeeper</li>
    <li>zkfuse</li>
    <li>zookeeper</li>
    <li>zookeeper_mt</li>
  </ul>
</details>

SuperSandro2000 commented:
Result of `nixpkgs-review pr 104889` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>perl530Packages.NetZooKeeper</li>
    <li>perl532Packages.NetZooKeeper</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>haskellPackages.hzk</li>
    <li>zookeeper</li>
    <li>zookeeper_mt</li>
  </ul>
</details>

ztzg commented:
Hi @SuperSandro2000,

Thank you for the review.

> 2 packages failed to build:
[…]
> Warning (mostly harmless): No library found for -lzookeeper_mt

Hmm…  I'm guessing I downed a Chesterton fence too many when I removed the `-L...` from `NIX_CFLAGS_LINK`.  I have now re-added it, but have no way to Darwin machine at hand, and thus no way to reproduce/test.

stigtsp commented:
@GrahamcOfBorg build perl530Packages.NetZooKeeper
@GrahamcOfBorg build perl532Packages.NetZooKeeper

stigtsp commented:
I'm (also) getting test failures on `darwin x68` for `perl530Packages.NetZooKeeper`, however `perl532Packages.NetZooKeeper` builds/tests OK

```
builder for '/nix/store/xhq0pasg89xjbmalv1mfhs7vscqky927-perl5.30.3-Net-ZooKeeper-0.41.drv' failed with exit code 2; last 10 log lines:
  t/30_connect.t  (Wstat: 512 Tests: 29 Failed: 2)
    Failed tests:  2, 22
    Non-zero exit status: 2
  t/45_class.t    (Wstat: 256 Tests: 47 Failed: 1)
    Failed test:  26
    Non-zero exit status: 1
  Files=11, Tests=445,  2 wallclock secs ( 0.07 usr  0.02 sys +  0.70 cusr  0.49 csys =  1.28 CPU)
  Result: FAIL
  Failed 3/11 test programs. 65/445 subtests failed.
  make: *** [Makefile:1034: test_dynamic] Error 255
cannot build derivation '/nix/store/40wkd932sr14bnlpcsn5hmv60j5w1p70-env.drv': 1 dependencies couldn't be built
[0 built (1 failed)]
error: build of '/nix/store/40wkd932sr14bnlpcsn5hmv60j5w1p70-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/104889
1 package failed to build:
perl530Packages.NetZooKeeper

4 packages built:
haskellPackages.hzk perl532Packages.NetZooKeeper zookeeper zookeeper_mt
```

ztzg commented:
Hi @stigtsp,

Thank you for the review.

> I'm (also) getting test failures on `darwin x68` for `perl530Packages.NetZooKeeper`, however `perl532Packages.NetZooKeeper` builds/tests OK
[…]
> Also, ofBorg builds [fail](https://github.com/NixOS/nixpkgs/pull/104889/checks?check_run_id=1459692689) on a test on `linux x68` and `linux aarch64`:

Yeah.  The remaining issues are a priori not `macOS` related, but just tests being flaky.

Given that the test suite is run in an environment where no ZooKeeper server is available, most tests are in fact being skipped.  The ZooKeeper library being highly asynchronous, I suspect some of the remaining paths suffer from races.  (The failing `t/35_log.t` above, for example, seems to be grepping the log before the connecting thread has had a chance to observe the TCP connection failure.)

I am consequently considering completely disabling the test suite.

> Upstream for [perlPackages.NetZooKeeper](https://github.com/mark-5/p5-net-zookeeper) has some recent patches in master for zookeeper 3.6, this might be an option.

Indeed.  I now realize that the situation with the CPAN Perl binding is not ideal, having diverged from ZooKeeper upstream.  I have [added a note and asked about upcoming releases](https://github.com/mark-5/p5-net-zookeeper/issues/12#issue-755126050); perhaps Mark has some suggestions.

In the meantime, I have pushed a new iteration with fetches the current tip of his `master` branch, and while it makes integration easier, I don't have high hopes wrt. tests—as I just saw it (sporadically) fail `t/35_log.t` on my workstation.

Cheers, -D

stigtsp commented:
> In the meantime, I have pushed a new iteration with fetches the current tip of his `master` branch, and while it makes integration easier, I don't have high hopes wrt. tests—as I just saw it (sporadically) fail `t/35_log.t` on my workstation.

Hey @ztzg  !

- Something like `preCheck = "rm t/35_log.t";` is imho preferable to `doCheck=false;`, if possible, so we have a way of catching issues when there are updates to perl or dependencies.

- Fetching from `https://github.com/mark-5/p5-net-zookeeper` looks legit to me, it is referenced from `https://metacpan.org/pod/Net::ZooKeeper`.

- It would be great to include tests for `perlPackages.NetZooKeeper` against a zookeeper instance in nixos/tests/zookeeper.nix too, I'm not familiar with ZooKeeper but have created a  test code snippet  that can be added if you like, otherwise I can do a separate PR once this is one is merged.

```
let
  perlEnv = pkgs.perl.withPackages (p: [p.NetZooKeeper]);

...

    server.wait_until_succeeds(
        "${perlEnv}/bin/perl -E 'use Net::ZooKeeper qw(:acls); $z=Net::ZooKeeper->new(q(localhost:2181));$z->create(qw(/perl foo acl), ZOO_OPEN_ACL_UNSAFE) || die $z->get_error()'"
    )


```

ztzg commented:
Hi @stigtsp,

Excellent; I'll respin this PR with your suggestions ASAP.  Thanks!

Cheers, -D

ztzg commented:
> * Something like `preCheck = "rm t/35_log.t";` is imho preferable to `doCheck=false;`, if possible, so we have a way of catching issues when there are updates to perl or dependencies.

Done.  (Well, removed `t/35_log.t`—which is is the one I saw failing.  But you also observed `t/30_connect.t` and `t/45_class.t` failures on Darwin; I did not do anything about those.)

> * Fetching from `https://github.com/mark-5/p5-net-zookeeper` looks legit to me, it is referenced from `https://metacpan.org/pod/Net::ZooKeeper`.

Okay, I left that one in, labeled with version "0.42pre."

> * It would be great to include tests for `perlPackages.NetZooKeeper` against a zookeeper instance in nixos/tests/zookeeper.nix too, I'm not familiar with ZooKeeper but have created a  test code snippet  that can be added if you like, otherwise I can do a separate PR once this is one is merged.

Done.  (I have also added a `get`.)

I have tried using `.succeed()` instead of `.wait_until_succeeds()`, but that led to spurious failures with -4, `ZCONNECTIONLOSS`.  (That Perl binding is a bit raw, and does not wait for the connection to be established before it tries and fires the `create` RPC.)  A `sleep(1)`  would help, but is undesirable for other reasons.  (Really, the only truly reasonable solution would be to improve the module.)  Oh well…

stigtsp commented:
@GrahamcOfBorg test zookeeper

stigtsp commented:
@GrahamcOfBorg build perlPackages.ZooKeeper perldevelPackages.ZooKeeper

stigtsp commented:
@GrahamcOfBorg build perPackages.NetZooKeeper perldevelPackages.NetZooKeeper

stigtsp reviewed (commented):
LGTM on `perlPackages.NetZooKeeper` and `nixos/tests/zookeeper.nix`

aanderse commented:
@stigtsp sorry, not familiar with this program at all. Can't really review it properly. :disappointed:

SuperSandro2000 commented:
Result of `nixpkgs-review pr 104889` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>haskellPackages.hzk</li>
    <li>perl530Packages.NetZooKeeper</li>
    <li>perl532Packages.NetZooKeeper</li>
    <li>zkfuse</li>
    <li>zookeeper</li>
    <li>zookeeper_mt</li>
  </ul>
</details>

stigtsp commented:
> @stigtsp sorry, not familiar with this program at all. Can't really review it properly. disappointed

Likewise, however changes look legit and tests pass. I'm inclined to merge unless there are objections.

ztzg commented:
@stigtsp: I don't have any outstanding items, so +1 from me.

(Of course, I may be a bit biased :)  But I provide ZooKeeper support professionally, am a committer on the Apache project, and am planning to stick around for a while.)

stigtsp commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -15705,18 +15705,26 @@ let
+    # Most tests are skipped as no server is available in the sandbox.
+    # `t/35_log.t` seems to suffer from a race condition; remove it.  See
+    # https://github.com/NixOS/nixpkgs/pull/104889#issuecomment-737144513
+    preCheck = "rm t/35_log.t";
```
One more thing: As @SuperSandro2000 found, some tests fail on darwin x64, we can remove the files containing those tests.

```suggestion
    preCheck = ''
      rm t/35_log.t
    '' + stdenv.lib.optionalString stdenv.isDarwin ''
      rm t/30_connect.t
      rm t/45_class.t
    '';
```

ztzg replied:
Incorporated!

stigtsp commented:
@GrahamcOfBorg build perl530Packages.NetZooKeeper perl532Packages.NetZooKeeper perldevelPackages.NetZooKeeper
@GrahamcOfBorg build zkfuse zookeeper zookeeper_mt haskellPackages.hzk
@GrahamcOfBorg test zookeeper

stigtsp reviewed (approved):
Looks good to me. Thank you for the contributions, and welcome to the NixOS project!

- Source are legit, and other changes to derivations look ok
- Did some basic `zkCli.sh` tests of the zookeeper nixos module in a nixos-shell
- `perlPackages.NetZooKeeper` is OK

ztzg commented:
Thanks!
